dagrunner/plugin_framework.py

```diff
@@ -59,6 +64,105 @@ def __call__(self, *args, **kwargs):
         return subprocess.run(*args, **kwargs, shell=True, check=True)
 
 
+def _stage_to_dir(*args, staging_dir, verbose=False):
+    """
+    Copy input filepaths to a staging area and update paths.
+
+    Hard link copies are preferred (same host) and physical copies are made otherwise.
+    File name, size and modification time are used to evaluate if the destination file
+    exists already (matching criteria of rsync).  If exists already, skip the copy.
+    Staged files are named: `<modification-time>_<file-size>_<filename>` to avoid
+    collision with identically names files.
+    """
+    os.makedirs(staging_dir, exist_ok=True)
+    args = list(args)
+    for ind, arg in enumerate(args):
+        host, fpath = None, arg
+        if ":" in arg:
+            host, fpath = arg.split(":")
+
+        if host:
+            source_mtime_size = subprocess.run(
+                ["ssh", host, "stat", "-c", "%Y_%s", fpath],
+                check=True,
+                text=True,
+                capture_output=True,
+            ).stdout.strip()
+        else:
+            source_mtime_size = (
+                f"{int(os.path.getmtime(fpath))}_{os.path.getsize(fpath)}"
+            )
+
+        target = os.path.join(
+            staging_dir, f"{source_mtime_size}_{os.path.basename(fpath)}"
+        )
+        if not os.path.exists(target):
+            if host:
+                rsync_command = ["scp", "-p", f"{host}:{fpath}", target]
+                subprocess.run(
+                    rsync_command, check=True, text=True, capture_output=True
+                )
```

Comment:
In an ideal network world connections always connect. In reality there are connection attempt failures and connections get dropped. I wondered if this should be a short (maybe 2 or 3 attempts) loop that will retry in the event of a failure. This could greatly improve reliability of doing the scp command.